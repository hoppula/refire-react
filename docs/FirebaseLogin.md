# FirebaseLogin

> Higher order component for Firebase login with e-mail and password

## Options

**validator** *(state) => Boolean*

Optional. Custom validator function.

```js
import { FirebaseLogin } from 'refire-react'
class LoginComponent extends Component {
  render() {
    const {
      email,
      password,
      submit,
      updateEmail,
      updatePassword,
      validInput,
      error,
      processing,
      completed
    } = this.props
  }
}
export default FirebaseLogin({
  validator: optionalCustomValidatorFn
})(LoginComponent)
```
