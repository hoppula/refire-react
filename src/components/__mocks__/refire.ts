import {Â firebaseActions as actions } from "refire"
export { syncFirebase, firebaseToProps, firebaseReducer } from "refire"
export const firebaseActions = {
  ...actions,
  passwordLogin: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  createUser: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  resetPassword: jest.fn((email) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  write: jest.fn((method, path, value, ownProps) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  clearWriteErrors: jest.fn((path) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
}
