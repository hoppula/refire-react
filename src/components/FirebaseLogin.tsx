import * as React from "react"
import { Component } from "react"
import * as ReactRedux from "react-redux"
import { connect } from "react-redux"
import { firebaseActions, firebaseToProps } from "refire"

const { clearLoginError, passwordLogin } = firebaseActions

interface ValidatorState {
  [key: string]: any
}

const defaultValidator = (state: ValidatorState = {}): boolean => {
  return Object.keys(state).every((field) => {
    return !!state[field]
  })
}

export interface Status {
  errors: {
    login: string
  },
  processing: {
    login: boolean
  },
  completed: {
    login: boolean
  }
}

export interface ExtraProps {
  email?: string,
  password?: string,
  submit?: (event: React.FormEvent<HTMLElement>) => void,
  updateEmail?: (event: React.FormEvent<HTMLInputElement>) => void,
  updatePassword?: (event: React.FormEvent<HTMLInputElement>) => void,
  validInput?: boolean,
  error?: string,
  processing?: boolean,
  completed?: boolean
}

export interface LoginOptions {
  validator?: (state: LoginState) => boolean
}

export interface LoginProps {
  readonly _status: Status,
  readonly dispatch?: ReactRedux.Dispatch<{}>
}

export interface LoginState {
  readonly email?: string,
  readonly password?: string
}

export default function(options: LoginOptions = {}) {

  const { validator = defaultValidator } = options

  return function <P>(
    WrappedComponent: ReactRedux.Component<P>
  ) {
    class FirebaseLogin extends Component<LoginProps & P, LoginState> {

      constructor(props: LoginProps & P) {
        super(props)
        this.state = {
          email: null,
          password: null
        }
      }

      submit = (event: React.FormEvent<HTMLElement>) => {
        event.preventDefault()
        this.props.dispatch(
          passwordLogin(this.state.email, this.state.password)
        ).catch(() => {})
      }

      updateEmail = (event: React.FormEvent<HTMLInputElement>) => {
        const { errors: { login: error } } = this.props._status
        if (error) {
          this.props.dispatch(clearLoginError())
        }
        this.setState({ email: event.currentTarget.value })
      }

      updatePassword = (event: React.FormEvent<HTMLInputElement>) => {
        const { errors: { login: error } } = this.props._status
        if (error) {
          this.props.dispatch(clearLoginError())
        }
        this.setState({ password: event.currentTarget.value })
      }

      render() {
        const validInput: boolean = validator(this.state)
        const {
          errors: { login: error },
          processing: { login: processing },
          completed: { login: completed }
        } = this.props._status

        const extraProps = {
          email: this.state.email,
          password: this.state.password,
          submit: this.submit,
          updateEmail: this.updateEmail,
          updatePassword: this.updatePassword,
          validInput,
          error,
          processing,
          completed
        }

        return (
          <WrappedComponent
            { ...this.props }
            { ...extraProps }
          />
        )
      }

    }

    return connect<LoginProps, void, P>(
      firebaseToProps(["_status"])
    )(FirebaseLogin)
  }
}
