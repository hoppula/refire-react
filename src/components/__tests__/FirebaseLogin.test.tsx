import * as React from "react"
import { Provider } from "react-redux"
import * as renderer from "react-test-renderer"
import { applyMiddleware, compose, createStore } from "redux"
import thunk from "redux-thunk"

import { firebaseActions } from "refire"
// passwordLogin action is mocked in ../__mocks__/refire.ts
const { passwordLogin } = firebaseActions

import FirebaseLogin from "../FirebaseLogin"

function createEmptyStore() {
  return compose(
    applyMiddleware(thunk)
  )(createStore)((state, action) => {
    return {
      firebase: {
        initialValuesReceived: true,
        errors: {
          login: null
        },
        processing: {
          login: false
        },
        completed: {
          login: false
        }
      }
    }
  })
}

describe("FirebaseLogin", () => {
  it("updateEmail updates email", () => {

    interface EmailProps {
      email?: string,
      updateEmail?: (event: any) => void,
    }
    function Email({ email, updateEmail }: EmailProps) {
      return (
        <div onClick={ updateEmail }>
          { email }
        </div>
      )
    }

    const LoginComponent = FirebaseLogin({})(Email)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <LoginComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({currentTarget: {value: "test@test.com"}})

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })


  it("updatePassword updates password", () => {

    interface PasswordProps {
      password?: string,
      updatePassword?: (event: any) => void,
    }
    function Password({ password, updatePassword }: PasswordProps) {
      return (
        <div onClick={ updatePassword }>
          { password }
        </div>
      )
    }

    const LoginComponent = FirebaseLogin({})(Password)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <LoginComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({currentTarget: {value: "password"}})

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("submit calls dispatch prop", () => {

    interface SubmitProps {
      dispatch?: () => void,
      submit?: (event: any) => void
    }
    function Submit({ submit }: SubmitProps) {
      return (
        <button onClick={ submit }>
          Submit
        </button>
      )
    }

    const LoginComponent = FirebaseLogin({})(Submit)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <LoginComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({ preventDefault() {} })
    expect(passwordLogin.mock.calls.length).toBe(1)
  })

  it("validator functions properly")
  it("errors are generated")
  it("processing prop gets updated")
  it("completed prop gets updated")

})




