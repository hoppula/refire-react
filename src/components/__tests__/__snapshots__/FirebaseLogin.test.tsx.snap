// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseLogin submit calls dispatch prop 1`] = `
<button
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`FirebaseLogin updateEmail updates email 1`] = `
<div
  onClick={[Function]}
/>
`;

exports[`FirebaseLogin updateEmail updates email 2`] = `
<div
  onClick={[Function]}
>
  test@test.com
</div>
`;

exports[`FirebaseLogin updatePassword updates password 1`] = `
<div
  onClick={[Function]}
/>
`;

exports[`FirebaseLogin updatePassword updates password 2`] = `
<div
  onClick={[Function]}
>
  password
</div>
`;
